### PR TITLE
Initializing the Parity hbbft Engine using the TOML config file

### PR DESCRIPTION
**Pull request by dforsten**
Temporary measure until the data necessary to initialize the hbbft NetworkInfo struct can be provided by smart contracts.

Further changes of note:
* Implemented TOML config file generator for hbbft validator networks of arbitrary size.
* Eliminated dependency on hbbft from ethcore
* More elegant decoding of hbbft batch transactions

### CODE REVIEW

File: ethcore/hbbft_engine/hbbft_config_generator/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	#[test]
+	fn test_network_info_serde() {
+		let mut rng = rand::thread_rng();
+		let net_infos = NetworkInfo::generate_map(0..1usize, &mut rng).unwrap();
+		let net_info = net_infos.get(&0).unwrap();
+
+		let toml_string = toml::to_string(&to_toml(&net_info)).unwrap();
+
+		// For debugging toml output:
+		//println!("{}", toml_string);
```
Comment (afck) [inline ethcore/hbbft_engine/hbbft_config_generator/src/main.rs]: Was that comment left over by accident?
If not, I'd add a space after "`//`".

File: ethcore/hbbft_engine/src/hbbft_engine.rs [hunk 2]
```diff
@@ -44,19 +46,44 @@ impl HoneyBadgerBFT {
...
 		if let Some(ref weak) = *self.client.read() {
 			if let Some(client) = weak.upgrade() {
 				// TODO: Retrieve the information to build a node-specific NetworkInfo
 				//       struct from the chain spec and from contracts.
-				let net_info = client.net_info().unwrap();
+				let options = client
+					.hbbft_options()
+					.expect("hbbft options are expected to exist");
+				let net_info = HoneyBadgerBFT::new_network_info(options)
+					.expect("hbbft parameter deserialization is expected to succeed");
```
Comment (afck) [inline ethcore/hbbft_engine/src/hbbft_engine.rs]: Maybe this should tell the user more explicitly that their config file is wrong.

File: ethcore/hbbft_engine/src/hbbft_engine.rs [hunk 3]
```diff
@@ -142,9 +159,13 @@ impl HoneyBadgerBFT {
...
-						let net_info = client.net_info().unwrap();
-						for peer_id in net_info.all_ids().filter(|p| p != &net_info.our_id()) {
-							client.send_consensus_message(ser.clone(), *peer_id);
+
+						if let Some(ref net_info) = *self.network_info.read() {
+							for peer_id in net_info.all_ids().filter(|p| p != &net_info.our_id()) {
+								client.send_consensus_message(ser.clone(), *peer_id);
+							}
+						} else {
+							panic!("Network Info needs to be initialized when dispatching hbbft consensus messages");
```
Comment (afck) [inline ethcore/hbbft_engine/src/hbbft_engine.rs]: Maybe shorter:
```rust
self.network_info.read().expect("Network Info needs to be…")
```

File: ethcore/src/miner/mod.rs [hunk 4]
```diff
@@ -51,6 +51,22 @@ use client::{
...
+/// TODO: Remove once contracts supply the necessary data.
+#[derive(Deserialize, Default, Clone)]
+pub struct HbbftOptions {
+	/// The node id of this validator
+	pub hbbft_our_id: String,
+	/// Threshold Cryptography secret share.
+	pub hbbft_secret_share: String,
+	/// The secret key of this hbbft validator.
+	pub hbbft_secret_key: String,
+	/// Threshold Cryptography public keys.
```
Comment (afck) [inline ethcore/src/miner/mod.rs]: Or "`public key shares`", to be even less ambiguous?

File: ethcore/hbbft_engine/hbbft_config_generator/src/main.rs (hunk 1 above)
Comment (dforsten) [reply ethcore/hbbft_engine/hbbft_config_generator/src/main.rs]: Removed!

File: ethcore/hbbft_engine/src/hbbft_engine.rs (hunk 2 above)
Comment (dforsten) [reply ethcore/hbbft_engine/src/hbbft_engine.rs]: Fixed

File: ethcore/hbbft_engine/src/hbbft_engine.rs (hunk 3 above)
Comment (dforsten) [reply ethcore/hbbft_engine/src/hbbft_engine.rs]: Shortened

File: ethcore/src/miner/mod.rs (hunk 4 above)
Comment (dforsten) [reply ethcore/src/miner/mod.rs]: Corrected

**Pull request merged**